### PR TITLE
Add Bazel integration

### PR DESCRIPTION
**Pull request by nak3**
Fixes https://github.com/google/autocxx/issues/478

This patch adds Bazel Integration.

It can build:
* `autocxx-*` libs/binaries - please see `BUILD` in the top directory.
* `demo` - please see `demo/BUILD`.

,which follows the structure of cxx project. So CI also built the demo app.

The commits contain a tons of file change but most of them are autogen files.
The autogen and non-autogen files are separated so please refer to the commit message.

> It's a good idea to open an issue first for discussion.

- [X] Tests pass
- [X] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by nak3** [comment]
Though I opened this PR I know I used some "hacky" way in `demo/BUILD`
so I don't mind if this PR was refused or keep it open until the satisfied solution is found.

File: BUILD [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+rust_library(
+    name = "autocxx-parser",
+    srcs = glob(["parser/src/**/*.rs"]),
+    edition = "2021",
+    deps = [
+        "//third-party:log",
+        "//third-party:proc-macro2",
+        "//third-party:quote",
+        "//third-party:serde",
+        "//third-party:thiserror",
```
Comment (adetaylor) [inline BUILD:42]: Thanks for this PR.

However, I hadn't realized that bazel support would require hard-coding all these dependencies. That's not OK with me - the `Cargo.toml` files need to be the sole source of truth - can you instead add a script or tool to generate these `BUILD` files based on the `Cargo.toml`?

More generally, I'm concerned that the project structure is not static enough to maintain a duplicate set of build instructions.

File: WORKSPACE [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    sha256 = "aaaa4b9591a5dad8d8907ae2dbe6e0eb49e6314946ce4c7149241648e56a1277",
+    urls = ["https://github.com/bazelbuild/rules_rust/releases/download/0.16.1/rules_rust-v0.16.1.tar.gz"],
+)
+
+load("@rules_rust//rust:repositories.bzl", "rules_rust_dependencies", "rust_register_toolchains")
+
+rules_rust_dependencies()
+
+rust_register_toolchains(
+    versions = ["1.66.0"],
```
Comment (adetaylor) [inline WORKSPACE:16]: Is this dependent on the CI environment?

File: BUILD (hunk 1 above)
Comment (nak3) [reply BUILD]: Thank you for the review. Yes, it was possible to drop the hard-coding deps in `BUILD` by [all_crate_deps](https://bazelbuild.github.io/rules_rust/crate_universe.html#all_crate_deps). I missed this utility.

File: WORKSPACE (hunk 2 above)
Comment (nak3) [reply WORKSPACE]: Ah, it is not necessary. It should be alright for `stable` and `nigthly` which will be registered as per http://bazelbuild.github.io/rules_rust/.

File: third-party/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+[workspace]
+[package]
+name = "third-party"
```
[the remark this answers is not in the document]
Comment (nak3) [reply third-party/Cargo.toml]: Thank you. I could drop these files. Now the generated files are only under `third-party/bazel/*` generated by `bazel run //third-party:vendor`.

[the remark this answers is not in the document]
Comment (adetaylor) [reply third-party/Cargo.toml]: Sorry for the lack of comment here - I've been travelling a bit lately.

I'm still not OK with these being checked in. I want to be able to upgrade dependencies by changing only `Cargo.toml` files - then everything else needs to be autogenerated.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -153,6 +153,28 @@ jobs:
...
+    # cxx repo disables to run the bazel job due to GitHub Actions's remote issue.
+    # It might be better to disable it here as well.
+    # See: https://github.com/dtolnay/cxx/commit/42c1ee806f186da58d1bddc8efea7d071a1c54af
+    #
+    # if: github.event_name != 'pull_request'
+    timeout-minutes: 45
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install Bazel
+        run: |
```
Comment (louwers) [inline .github/workflows/ci.yml:168]: You should use `.bazelversion`. Bazelisk is already installed on the Ubuntu image. You don't need to install anything.

Bazel 6.0.0 is old. A newer version should be used.

File: WORKSPACE [hunk 5]
```diff
@@ -0,0 +1,23 @@
+workspace(name = "autocxx.rs")
```
Comment (louwers) [inline WORKSPACE:1]: `WORKSPACE` is deprecated and disabled by default in the most recent Bazel version.

I would use a `MODULE.bazel` file if possible.

File: third-party/bazel/BUILD.addr2line-0.17.0.bazel [hunk 6]
```diff
@@ -0,0 +1,43 @@
+###############################################################################
+# @generated
+# DO NOT MODIFY: This file is auto-generated by a crate_universe tool. To
```
Comment (louwers) [inline third-party/bazel/BUILD.addr2line-0.17.0.bazel:3]: Generated files should not be checked in.

**Comment by nak3** [comment]
Hi @louwers
Thank you for reviewing!
Unfortunately, I have recently been moving away from Bazel, so I am in a difficult position to address the issues you pointed out. Therefore, I plan to close this PR for now.
Please feel free to take this PR and create a new one if you’d like.

**Pull request closed without merging**